### PR TITLE
Proxy for credentials

### PR DESCRIPTION
**Pull request by erwannT**
This pull request adds a new section in "Configuration" form in order to configure an internet proxy for SSH connection.
Based on a feature of JSCH, an HTTP, SOCKS4 or SOCKS5 proxy  can be configured. A type of proxy, url and port of proxy server, and optionally username and password can be specified.

### CODE REVIEW

File: src/main/java/jenkins/plugins/publish_over_ssh/BapSshHostConfiguration.java [hunk 1]
```diff
@@ -42,51 +46,71 @@
...
 public class BapSshHostConfiguration extends BPHostConfiguration<BapSshClient, BapSshCommonConfiguration>
-                                                                                        implements Describable<BapSshHostConfiguration> {
+        implements Describable<BapSshHostConfiguration> {
 
     private static final long serialVersionUID = 1L;
     public static final int DEFAULT_PORT = 22;
     public static final int DEFAULT_TIMEOUT = 300000;
     public static final String CONFIG_KEY_PREFERRED_AUTHENTICATIONS = "PreferredAuthentications";
     private static final Log LOG = LogFactory.getLog(BapSshHostConfiguration.class);
+    public static final String HTTP_PROXY_TYPE = "http";
```
Comment (slide) [inline src/main/java/jenkins/plugins/publish_over_ssh/BapSshHostConfiguration.java]: Are these not already defined in the jsch library somewhere?

File: src/main/java/jenkins/plugins/publish_over_ssh/BapSshHostConfiguration.java [hunk 2]
```diff
@@ -42,51 +46,71 @@
...
+    private String proxyType;
+    private String proxyHost;
+    private int proxyPort;
+    private String proxyUser;
+    private String proxyPassword;
+
     // CSOFF: ParameterNumberCheck
     @SuppressWarnings("PMD.ExcessiveParameterList") // DBC for you!
     @DataBoundConstructor
     public BapSshHostConfiguration(final String name, final String hostname, final String username, final String encryptedPassword,
```
Comment (slide) [inline src/main/java/jenkins/plugins/publish_over_ssh/BapSshHostConfiguration.java]: Please use DataBoundSetters to add new fields instead of adding more parameters to the constructor

File: src/main/java/jenkins/plugins/publish_over_ssh/BapSshHostConfiguration.java [hunk 3]
```diff
@@ -42,51 +46,71 @@
...
+        this.proxyUser = proxyUser;
+    }
+
+    public int getTimeout() {
+        return timeout;
     }
 
-    public int getTimeout() { return timeout; }
-    public void setTimeout(final int timeout) { this.timeout = timeout; }
+    public void setTimeout(final int timeout) {
```
Comment (slide) [inline src/main/java/jenkins/plugins/publish_over_ssh/BapSshHostConfiguration.java]: Why the whitespace changes?

File: src/main/java/jenkins/plugins/publish_over_ssh/BapSshHostConfiguration.java (hunk 2 above)
Comment (mathieu-pousse) [inline src/main/java/jenkins/plugins/publish_over_ssh/BapSshHostConfiguration.java]: DataBoundSetters have been introduced in 1.535 (https://groups.google.com/d/msg/jenkinsci-dev/58-DEvuJZWI/5QrxBZRFJ6IJ) but this project depends on 1.532

maybe the upgrade should be done on a separate PR to keep it clear

Comment (slide) [inline src/main/java/jenkins/plugins/publish_over_ssh/BapSshHostConfiguration.java]: Yeah, it would be nice to update it.

File: src/main/java/jenkins/plugins/publish_over_ssh/BapSshHostConfiguration.java (hunk 1 above)
Comment (mathieu-pousse) [inline src/main/java/jenkins/plugins/publish_over_ssh/BapSshHostConfiguration.java]: does not seem to be definied neither in `com.jcraft.jsch.Proxy` or sibling classes, nor in `java.net.Proxy`

**Comment by mathieu-pousse** [comment]
I have upgraded to 1.554.3 the oldest LTS that allows the DataBoundSetter. I have cleanup the code accordingly.

Thanks in advance

**Comment by slide** [comment]
My apologies, I wasn't meaning for you to decompose the whole constructor into DataBoundSetters, just the new field(s) that you were adding. This would maintain the constructor that was already there, but allow for adding new things to the configuration.

**Comment by mathieu-pousse** [comment]
ok for you if I revert the constructor only ?

**Comment by slide** [comment]
Yeah, that should be ok

**Comment by mathieu-pousse** [comment]
should be done, i've also squashed everything to one commit

**Comment by mathieu-pousse** [comment]
nice, any plan for the release ?

thanks
